vck/src/commonMain/kotlin/at/asitplus/wallet/lib/cbor/CoseService.kt

```diff
@@ -134,65 +154,76 @@ object CoseUtils {
-            }.firstNotNullOf { coseKey ->
-                verifyCoseSignature(coseSigned, coseKey, externalAad, detachedPayload).getOrNull()
-            }
+fun interface VerifyCoseSignatureFun<P> {
+    suspend operator fun invoke(
+        coseSigned: CoseSigned<P>,
+        externalAad: ByteArray,
+        detachedPayload: ByteArray?,
+    ): KmmResult<Verifier.Success>
+}
+
+class VerifyCoseSignature<P : Any>(
+    val verifyCoseSignature: VerifyCoseSignatureWithKeyFun<P> = VerifyCoseSignatureWithKey<P>(),
+    /** Need to implement if valid keys for CoseSigned are transported somehow out-of-band, e.g. provided by a trust store */
+    val publicKeyLookup: PublicCoseKeyLookup = PublicCoseKeyLookup { null }
+) : VerifyCoseSignatureFun<P> {
+    override suspend fun invoke(
+        coseSigned: CoseSigned<P>,
+        externalAad: ByteArray,
+        detachedPayload: ByteArray?,
+    ) = catching {
+        coseSigned.loadPublicKeys().also {
+            Napier.d("Public keys available: ${it.size}")
+        }.firstNotNullOf { coseKey ->
+            verifyCoseSignature(coseSigned, coseKey, externalAad, detachedPayload).getOrNull()
         }
     }
 
-    fun CoseSigned<*>.loadPublicKeys(
-        publicKeyLookup: PublicCoseKeyLookup = { null },
-    ): Set<CoseKey> = (protectedHeader.publicKey ?: unprotectedHeader?.publicKey)?.let { setOf(it) }
-        ?: publicKeyLookup(this) ?: setOf()
+    suspend fun CoseSigned<*>.loadPublicKeys(): Set<CoseKey> =
+        (protectedHeader.publicKey ?: unprotectedHeader?.publicKey)?.let { setOf(it) }
+            ?: publicKeyLookup(this) ?: setOf()
 }
 
-typealias VerifyCoseSignatureWithKeyFun<P> = (
-    coseSigned: CoseSigned<P>,
-    signer: CoseKey,
-    externalAad: ByteArray,
-    detachedPayload: ByteArray?,
-) -> KmmResult<Verifier.Success>
-
-object VerifyCoseSignatureWithKey {
-    operator fun <P : Any> invoke(
-        verifySignature: VerifySignatureFun = VerifySignature(),
-    ): VerifyCoseSignatureWithKeyFun<P> = { coseSigned, signer, externalAad, detachedPayload ->
-        catching {
-            val signatureInput = coseSigned.prepareCoseSignatureInput(externalAad, detachedPayload)
-                .also { Napier.d("verifyCose input is ${it.encodeToString(Base16())}") }
-            val algorithm = coseSigned.protectedHeader.algorithm
-                ?: throw IllegalArgumentException("Algorithm not specified")
-            val publicKey = signer.toCryptoPublicKey().getOrElse { ex ->
-                throw IllegalArgumentException("Signer not convertible", ex)
-                    .also { Napier.w("Could not convert signer to public key: $signer", ex) }
-            }
-            verifySignature(
-                signatureInput,
-                coseSigned.signature,
-                algorithm.algorithm,
-                publicKey
-            ).getOrThrow()
+fun interface VerifyCoseSignatureWithKeyFun<P> {
+    operator fun invoke(
```

Comment:
Considering we will be adding remote PQC capabilities to Signum, this should be suspending